File: app/src/androidTest/java/com/example/mobdev21_night_at_the_museum/presentation/intro/StreetViewTest.java
package com.example.mobdev21_night_at_the_museum.presentation.intro;


import static androidx.test.espresso.Espresso.onView;
import static androidx.test.espresso.action.ViewActions.click;
import static androidx.test.espresso.assertion.ViewAssertions.matches;
import static androidx.test.espresso.matcher.ViewMatchers.isDisplayed;
import static androidx.test.espresso.matcher.ViewMatchers.withId;
import static androidx.test.espresso.matcher.ViewMatchers.withParent;
import static androidx.test.espresso.matcher.ViewMatchers.withText;
import static org.hamcrest.Matchers.allOf;

import android.view.View;
import android.view.ViewGroup;
import android.view.ViewParent;

import androidx.test.espresso.ViewInteraction;
import androidx.test.ext.junit.rules.ActivityScenarioRule;
import androidx.test.ext.junit.runners.AndroidJUnit4;
import androidx.test.filters.LargeTest;

import com.example.mobdev21_night_at_the_museum.R;

import org.hamcrest.Description;
import org.hamcrest.Matcher;
import org.hamcrest.TypeSafeMatcher;
import org.hamcrest.core.IsInstanceOf;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;

@LargeTest
@RunWith(AndroidJUnit4.class)
public class StreetViewTest {

    @Rule
    public ActivityScenarioRule<CheckPrefActivity> mActivityScenarioRule =
            new ActivityScenarioRule<>(CheckPrefActivity.class);

    @Test
    public void streetViewTest() {
        ViewInteraction materialCardView = onView(
                allOf(withId(R.id.mcvHomeStreetViewViewAll),
                        childAtPosition(
                                childAtPosition(
                                        withId(R.id.cvHome),
                                        1),
                                4),
                        isDisplayed()));
        materialCardView.perform(click());

        ViewInteraction linearLayout = onView(
                allOf(withParent(withParent(withId(R.id.cvAllStreetView))),
                        isDisplayed()));
        linearLayout.check(matches(isDisplayed()));

        ViewInteraction imageView = onView(
                allOf(withId(R.id.ivHomeStreetViewThumbnail),
                        withParent(withParent(IsInstanceOf.<View>instanceOf(android.widget.LinearLayout.class))),
                        isDisplayed()));
        imageView.check(matches(isDisplayed()));

        ViewInteraction imageView2 = onView(
                allOf(withParent(withParent(IsInstanceOf.<View>instanceOf(android.widget.FrameLayout.class))),
                        isDisplayed()));
        imageView2.check(matches(isDisplayed()));

        ViewInteraction textView = onView(
                allOf(withText("KHÁM PHÁ"),
                        withParent(withParent(IsInstanceOf.<View>instanceOf(android.widget.FrameLayout.class))),
                        isDisplayed()));
        textView.check(matches(withText("KHÁM PHÁ")));


        ViewInteraction frameLayout = onView(
                allOf(childAtPosition(
                                allOf(withId(R.id.cvAllStreetView),
                                        childAtPosition(
                                                withId(R.id.flAllStreetViewContainer),
                                                0)),
                                0),
                        isDisplayed()));
        frameLayout.perform(click());

        ViewInteraction frameLayout2 = onView(
                allOf(childAtPosition(
                                allOf(withId(R.id.cvAllStreetView),
                                        childAtPosition(
                                                withId(R.id.flAllStreetViewContainer),
                                                0)),
                                0),
                        isDisplayed()));
        frameLayout2.perform(click());
    }

    private static Matcher<View> childAtPosition(
            final Matcher<View> parentMatcher, final int position) {

        return new TypeSafeMatcher<View>() {
            @Override
            public void describeTo(Description description) {
                description.appendText("Child at position " + position + " in parent ");
                parentMatcher.describeTo(description);
            }

            @Override
            public boolean matchesSafely(View view) {
                ViewParent parent = view.getParent();
                return parent instanceof ViewGroup && parentMatcher.matches(parent)
                        && view.equals(((ViewGroup) parent).getChildAt(position));
            }
        };
    }
}

File: app/src/androidTest/java/com/example/mobdev21_night_at_the_museum/presentation/intro/StreetViewTest.kt
package com.example.mobdev21_night_at_the_museum.presentation.intro


import android.view.View
import android.view.ViewGroup
import androidx.test.espresso.Espresso.onView
import androidx.test.espresso.action.ViewActions.*
import androidx.test.espresso.assertion.ViewAssertions.*
import androidx.test.espresso.matcher.ViewMatchers.*
import androidx.test.ext.junit.rules.ActivityScenarioRule
import androidx.test.ext.junit.runners.AndroidJUnit4
import androidx.test.filters.LargeTest
import com.example.mobdev21_night_at_the_museum.R
import org.hamcrest.Description
import org.hamcrest.Matcher
import org.hamcrest.Matchers.allOf
import org.hamcrest.Matchers.`is`
import org.hamcrest.TypeSafeMatcher
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith

@LargeTest
@RunWith(AndroidJUnit4::class)
class StreetViewTest {

    @Rule
    @JvmField
    var mActivityScenarioRule = ActivityScenarioRule(CheckPrefActivity::class.java)

    @Test
    fun streetViewTest() {
//        Thread.sleep(10000)
//        val materialCardView = onView(
//            allOf(
//                withId(R.id.mcvIntroductionSignIn),
//                childAtPosition(
//                    childAtPosition(
//                        withId(R.id.constIntroductionRoot),
//                        3
//                    ),
//                    0
//                ),
//                isDisplayed()
//            )
//        )
//        materialCardView.perform(click())
//        Thread.sleep(5000)
//        val appCompatEditText = onView(
//            allOf(
//                withId(R.id.etvLoginEmail),
//                childAtPosition(
//                    childAtPosition(
//                        withClassName(`is`("androidx.constraintlayout.widget.ConstraintLayout")),
//                        2
//                    ),
//                    0
//                ),
//                isDisplayed()
//            )
//        )
//        appCompatEditText.perform(replaceText("sonanhnguyen003@gmail.com"), closeSoftKeyboard())
//        Thread.sleep(5000)
//        val appCompatEditText2 = onView(
//            allOf(
//                withId(R.id.etvLoginPassword),
//                childAtPosition(
//                    childAtPosition(
//                        withClassName(`is`("android.widget.LinearLayout")),
//                        1
//                    ),
//                    0
//                ),
//                isDisplayed()
//            )
//        )
//        appCompatEditText2.perform(replaceText("12345678"), closeSoftKeyboard())
//        Thread.sleep(5000)
//        val materialCardView2 = onView(
//            allOf(
//                withId(R.id.btnLogin),
//                childAtPosition(
//                    childAtPosition(
//                        withClassName(`is`("androidx.constraintlayout.widget.ConstraintLayout")),
//                        2
//                    ),
//                    2
//                ),
//                isDisplayed()
//            )
//        )
//        materialCardView2.perform(click())
        Thread.sleep(5000)
        val materialCardView3 = onView(
            allOf(
                withId(R.id.mcvHomeStreetViewViewAll),
                childAtPosition(
                    childAtPosition(
                        withId(R.id.cvHome),
                        1
                    ),
                    4
                )
            )
        )
        materialCardView3.perform(scrollTo())
        Thread.sleep(5000);
        materialCardView3.perform(click())
        Thread.sleep(5000)
        val frameLayout = onView(
            allOf(
                childAtPosition(
                    allOf(
                        withId(R.id.cvAllStreetView),
                        childAtPosition(
                            withId(R.id.flAllStreetViewContainer),
                            0
                        )
                    ),
                    0
                ),
                isDisplayed()
            )
        )
        frameLayout.perform(click())
        Thread.sleep(5000)
        val appCompatImageView = onView(
            allOf(
                withId(R.id.ivRealMainSearch),
                childAtPosition(
                    allOf(
                        withId(R.id.constRealMainActionBar),
                        childAtPosition(
                            withId(R.id.ablRealMain),
                            0
                        )
                    ),
                    0
                ),
                isDisplayed()
            )
        )
        appCompatImageView.perform(click())
        Thread.sleep(5000)
        val appCompatImageView2 = onView(
            allOf(
                withId(R.id.ivRealMainSearch),
                childAtPosition(
                    allOf(
                        withId(R.id.constRealMainActionBar),
                        childAtPosition(
                            withId(R.id.ablRealMain),
                            0
                        )
                    ),
                    0
                ),
                isDisplayed()
            )
        )
        appCompatImageView2.perform(click())
        Thread.sleep(5000)
        val shapeableImageView = onView(
            allOf(
                withId(R.id.ivRealMainProfile),
                childAtPosition(
                    allOf(
                        withId(R.id.constRealMainActionBar),
                        childAtPosition(
                            withId(R.id.ablRealMain),
                            0
                        )
                    ),
                    2
                ),
                isDisplayed()
            )
        )
        shapeableImageView.perform(click())
        Thread.sleep(3000)
        val linearLayout = onView(
            allOf(
                withId(R.id.llProfileSettings),
                childAtPosition(
                    allOf(
                        withId(R.id.llProfileContent),
                        childAtPosition(
                            withId(R.id.flProfileBackground),
                            0
                        )
                    ),
                    5
                ),
                isDisplayed()
            )
        )
        linearLayout.perform(click())
        Thread.sleep(3000)
        val materialTextView = onView(
            allOf(
                withId(R.id.tvSettingLogOut), withText("Đăng xuất"),
                childAtPosition(
                    childAtPosition(
                        withId(R.id.nsvSettingFragment),
                        0
                    ),
                    25
                ),

            )
        )
        materialTextView.perform(scrollTo())
        materialTextView.perform(click())
    }

    private fun childAtPosition(
        parentMatcher: Matcher<View>, position: Int
    ): Matcher<View> {

        return object : TypeSafeMatcher<View>() {
            override fun describeTo(description: Description) {
                description.appendText("Child at position $position in parent ")
                parentMatcher.describeTo(description)
            }

            public override fun matchesSafely(view: View): Boolean {
                val parent = view.parent
                return parent is ViewGroup && parentMatcher.matches(parent)
                        && view == parent.getChildAt(position)
            }
        }
    }
}
